code/datums/components/defibrillator.dm

```diff
@@ -0,0 +1,382 @@
+ */
+/datum/component/defib/Initialize(robotic, cooldown = 5 SECONDS, speed_multiplier = 1, combat = FALSE, combat_defib_chance = 100, safe_by_default = TRUE, emp_proof = FALSE, emag_proof = FALSE, obj/item/actual_unit = null)
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.robotic = robotic
+	src.speed_multiplier = speed_multiplier
+	src.cooldown = cooldown
+	src.combat = combat
+	src.combat_defib_chance = combat_defib_chance
+	safety = safe_by_default
+	src.emp_proof = emp_proof
+	src.emag_proof = emag_proof
+
+	if(actual_unit)
+		actual_unit_uid = actual_unit.UID()
+
+	var/effect_target = isnull(actual_unit) ? parent : actual_unit
+
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK, PROC_REF(trigger_defib))
+	RegisterSignal(effect_target, COMSIG_ATOM_EMAG_ACT, PROC_REF(on_emag))
+	RegisterSignal(effect_target, COMSIG_ATOM_EMP_ACT, PROC_REF(on_emp))
+
+/**
+ * Get the "parent" that effects (emags, EMPs) should be applied onto.
+ */
+/datum/component/defib/proc/get_effect_target()
+	var/actual_unit = locateUID(actual_unit_uid)
+	if(!actual_unit)
+		return parent
+	return actual_unit
+
+/datum/component/defib/proc/on_emp(obj/item/unit)
+	SIGNAL_HANDLER  // COMSIG_ATOM_EMP_ACT
+	if(emp_proof)
+		return
+
+	if(safety)
+		safety = FALSE
+		unit.visible_message("<span class='notice'>[unit] beeps: Safety protocols disabled!</span>")
+		playsound(get_turf(unit), 'sound/machines/defib_saftyoff.ogg', 50, 0)
+	else
+		safety = TRUE
+		unit.visible_message("<span class='notice'>[unit] beeps: Safety protocols enabled!</span>")
+		playsound(get_turf(unit), 'sound/machines/defib_saftyon.ogg', 50, 0)
+
+/datum/component/defib/proc/on_emag(obj/item/unit, mob/user)
+	SIGNAL_HANDLER  // COMSIG_ATOM_EMAG_ACT
+	if(emag_proof)
+		return
+	if(safety)
+		safety = FALSE
+		if(user && !robotic)
+			to_chat(user, "<span class='warning'>You silently disable [unit]'s safety protocols with the card.")
+	else
+		safety = TRUE
+		if(user && !robotic)
+			to_chat(user, "<span class='notice'>You silently enable [unit]'s safety protocols with the card.")
+
+/datum/component/defib/proc/set_cooldown()
+	on_cooldown = TRUE
+	addtimer(CALLBACK(src, PROC_REF(end_cooldown)), cooldown)
+
+/datum/component/defib/proc/end_cooldown()
+	on_cooldown = FALSE
+	SEND_SIGNAL(parent, COMSIG_DEFIB_READY)
+
+/**
+ * Start the defibrillation process when triggered by a signal.
+ */
+/datum/component/defib/proc/trigger_defib(obj/item/paddles, mob/living/carbon/human/target, mob/living/user)
+	SIGNAL_HANDLER  // COMSIG_ITEM_ATTACK
+	// This is directly referenced in the attack chain, so it can't sleep or do-after. Pass the work off asynchronously.
+	INVOKE_ASYNC(src, PROC_REF(defibrillate), user, target)
+	return TRUE
+
+/**
+ * Perform a defibrillation.
+ */
+/datum/component/defib/proc/defibrillate(mob/living/user, mob/living/carbon/human/target)
```

Comment:
Pour one out for my homies, this proc is 200+ lines

Reply:
Hey, it's better than two 200+ line procs that are almost exactly the same (but different in like one or two ways).

Reply:
hahaha